The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7050784238/job/19192218735) [(plain)](https://github.com/rust-lang/rust/commit/f545e8963d01484a4b99967b846f7b502bd6754f/checks/19192218735/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=blyxxyz
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_22594336-6309-435f-8954-a74843528718
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=b507ccbcee8baa39f8f3dbba0f76dba9d0c87739
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_22594336-6309-435f-8954-a74843528718
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_22594336-6309-435f-8954-a74843528718
GITHUB_TRIGGERING_ACTOR=blyxxyz
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/118484/merge
GITHUB_WORKFLOW_SHA=b507ccbcee8baa39f8f3dbba0f76dba9d0c87739
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
 Documenting std v0.0.0 (/checkout/library/std)
error: `str` is both a module and a primitive type
  --> library/std/src/ffi/os_str.rs:37:72
   |
37 | /// `OsString` is to <code>&[OsStr]</code> as [`String`] is to <code>&[str]</code>: the former
   |                                                                        ^^^ ambiguous link
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`
help: to link to the module, prefix with `mod@`
   |
37 | /// `OsString` is to <code>&[OsStr]</code> as [`String`] is to <code>&[mod@str]</code>: the former
   |                                                                        ++++
help: to link to the primitive type, prefix with `prim@`
   |
37 | /// `OsString` is to <code>&[OsStr]</code> as [`String`] is to <code>&[prim@str]</code>: the former

error: `str` is both a module and a primitive type
  --> library/std/src/ffi/os_str.rs:71:59
   |
   |
71 | /// [`String`] and then [`String::push_str`] some <code>&[str]</code>
   |                                                           ^^^ ambiguous link
help: to link to the module, prefix with `mod@`
   |
   |
71 | /// [`String`] and then [`String::push_str`] some <code>&[mod@str]</code>
   |                                                           ++++
help: to link to the primitive type, prefix with `prim@`
   |
71 | /// [`String`] and then [`String::push_str`] some <code>&[prim@str]</code>

error: `str` is both a module and a primitive type
   --> library/std/src/ffi/os_str.rs:105:44
    |
    |
105 | /// `&OsStr` is to [`OsString`] as <code>&[str]</code> is to [`String`]: the
    |                                            ^^^ ambiguous link
help: to link to the module, prefix with `mod@`
    |
    |
105 | /// `&OsStr` is to [`OsString`] as <code>&[mod@str]</code> is to [`String`]: the
    |                                            ++++
help: to link to the primitive type, prefix with `prim@`
    |
105 | /// `&OsStr` is to [`OsString`] as <code>&[prim@str]</code> is to [`String`]: the

error: `str` is both a module and a primitive type
   --> library/std/src/ffi/os_str.rs:793:26
    |
    |
793 |     /// Yields a <code>&[str]</code> slice if the `OsStr` is valid Unicode.
    |                          ^^^ ambiguous link
help: to link to the module, prefix with `mod@`
    |
    |
793 |     /// Yields a <code>&[mod@str]</code> slice if the `OsStr` is valid Unicode.
    |                          ++++
help: to link to the primitive type, prefix with `prim@`
    |
793 |     /// Yields a <code>&[prim@str]</code> slice if the `OsStr` is valid Unicode.

error: `str` is both a module and a primitive type
   --> library/std/src/ffi/os_str.rs:813:47
    |
    |
813 |     /// Converts an `OsStr` to a <code>[Cow]<[str]></code>.
    |                                               ^^^ ambiguous link
help: to link to the module, prefix with `mod@`
    |
    |
813 |     /// Converts an `OsStr` to a <code>[Cow]<[mod@str]></code>.
    |                                               ++++
help: to link to the primitive type, prefix with `prim@`
    |
813 |     /// Converts an `OsStr` to a <code>[Cow]<[prim@str]></code>.

error: could not document `std`

Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type dylib --crate-type rlib --crate-name std library/std/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/doc --cfg 'feature="addr2line"' --cfg 'feature="backtrace"' --cfg 'feature="compiler-builtins-c"' --cfg 'feature="gimli-symbolize"' --cfg 'feature="miniz_oxide"' --cfg 'feature="object"' --cfg 'feature="panic_unwind"' --cfg 'feature="std_detect_dlsym_getauxval"' --cfg 'feature="std_detect_file_io"' -Zunstable-options --check-cfg 'cfg(feature, values("addr2line", "backtrace", "compiler-builtins-c", "compiler-builtins-mangled-names", "compiler-builtins-mem", "compiler-builtins-no-asm", "compiler-builtins-weak-intrinsics", "gimli-symbolize", "llvm-libunwind", "miniz_oxide", "object", "panic-unwind", "panic_immediate_abort", "panic_unwind", "profiler", "profiler_builtins", "std_detect_dlsym_getauxval", "std_detect_env_override", "std_detect_file_io", "system-llvm-libunwind"))' --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat -C metadata=c9d015852ec79f15 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/release/deps --extern 'priv:addr2line=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libaddr2line-b302fd601f7d0808.rmeta' --extern alloc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/liballoc-6c01484629166c1b.rmeta --extern 'priv:cfg_if=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libcfg_if-089e5122ea9a0137.rmeta' --extern 'priv:compiler_builtins=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libcompiler_builtins-a26ae7424b376de1.rmeta' --extern core=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libcore-a1aec6da431e1dc8.rmeta --extern 'priv:hashbrown=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libhashbrown-6d3d2c5457ccb055.rmeta' --extern libc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/liblibc-322cea99859723d1.rmeta --extern 'priv:miniz_oxide=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libminiz_oxide-38a88a85563bc2d8.rmeta' --extern 'priv:object=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libobject-219974afe5b5cce3.rmeta' --extern 'priv:panic_abort=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libpanic_abort-bf4fd79f73d0bedf.rmeta' --extern 'priv:panic_unwind=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libpanic_unwind-de1b3ce07eb74da8.rmeta' --extern 'priv:rustc_demangle=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/librustc_demangle-56da456d52f5766d.rmeta' --extern 'priv:std_detect=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libstd_detect-690ad1c975789f88.rmeta' --extern 'priv:unwind=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc/x86_64-unknown-linux-gnu/release/deps/libunwind-581bdfc3e73f9e37.rmeta' -Z unstable-options --cfg=bootstrap -Csymbol-mangling-version=legacy -Zunstable-options '--check-cfg=cfg(feature,values(any()))' -Zunstable-options '--check-cfg=cfg(bootstrap,values())' '--check-cfg=cfg(stdarch_intel_sde,values())' '--check-cfg=cfg(no_fp_fmt_parse,values())' '--check-cfg=cfg(no_global_oom_handling,values())' '--check-cfg=cfg(no_rc,values())' '--check-cfg=cfg(no_sync,values())' '--check-cfg=cfg(freebsd12,values())' '--check-cfg=cfg(freebsd13,values())' '--check-cfg=cfg(backtrace_in_libstd,values())' '--check-cfg=cfg(target_env,values("libnx"))' '--check-cfg=cfg(target_arch,values("spirv","nvptx","xtensa"))' --document-private-items --document-hidden-items -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.76.0-nightly (b507ccbce 2023-11-30)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")' '-Zcrate-attr=warn(rust_2018_idioms)' -Z unstable-options --resource-suffix 1.76.0 --markdown-css rust.css --markdown-no-toc --index-page /checkout/src/doc/index.md --cfg backtrace_in_libstd` (exit status: 1)
Build completed unsuccessfully in 0:00:16
  local time: Thu Nov 30 18:46:48 UTC 2023
  network time: Thu, 30 Nov 2023 18:46:48 GMT
##[error]Process completed with exit code 1.
```

</details>

--- a/library/std/src/ffi/os_str.rs
+++ b/library/std/src/ffi/os_str.rs
@@ -6,9 +6,10 @@ use crate::cmp;
 use crate::collections::TryReserveError;
 use crate::fmt;
 use crate::hash::{Hash, Hasher};
-use crate::ops;
+use crate::ops::{self, Range};
 use crate::rc::Rc;
-use crate::str::FromStr;
+use crate::slice;
+use crate::str::{from_utf8 as str_from_utf8, FromStr};
 use crate::sync::Arc;
 
 use crate::sys::os_str::{Buf, Slice};
@@ -961,6 +962,83 @@ impl OsStr {
     #[stable(feature = "os_str_bytes", since = "1.74.0")]
     pub fn as_encoded_bytes(&self) -> &[u8] {
         self.inner.as_encoded_bytes()
+    }
+
+    /// Takes a substring based on a range that corresponds to the return value of
+    /// [`OsStr::as_encoded_bytes`].
+    ///
+    /// The range's start and end must lie on valid `OsStr` boundaries.
+    /// A valid `OsStr` boundary is one of:
+    /// - The start of the string
+    /// - The end of the string
+    /// - Immediately before a valid non-empty UTF-8 substring
+    /// - Immediately after a valid non-empty UTF-8 substring
+    ///
+    /// # Panics
+    ///
+    /// Panics if `range` does not lie on valid `OsStr` boundaries or if it
+    /// exceeds the end of the string.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// #![feature(os_str_slice)]
+    ///
+    /// use std::ffi::OsStr;
+    ///
+    /// let os_str = OsStr::new("foo=bar");
+    /// let bytes = os_str.as_encoded_bytes();
+    /// if let Some(index) = bytes.iter().position(|b| *b == b'=') {
+    ///     let key = os_str.slice_encoded_bytes(..index);
+    ///     let value = os_str.slice_encoded_bytes(index + 1..);
+    ///     assert_eq!(key, "foo");
+    ///     assert_eq!(value, "bar");
+    /// }
+    /// ```
+    #[unstable(feature = "os_str_slice", issue = "118485")]
+    pub fn slice_encoded_bytes<R: ops::RangeBounds<usize>>(&self, range: R) -> &Self {
+        #[track_caller]
+        fn check_valid_boundary(bytes: &[u8], index: usize) {
+            if index == 0 || index == bytes.len() {
+                return;
+            }
+
+            // Fast path
+            if bytes[index - 1].is_ascii() || bytes[index].is_ascii() {
+                return;
+            }
+
+            let (before, after) = bytes.split_at(index);
+
+            // UTF-8 takes at most 4 bytes per codepoint, so we don't
+            // need to check more than that.
+            let after = after.get(..4).unwrap_or(after);
+            match str_from_utf8(after) {
+                Ok(_) => return,
+                Err(err) if err.valid_up_to() != 0 => return,
+                Err(_) => (),
+            }
+
+            for len in 2..=4.min(index) {
+                let before = &before[index - len..];
+                if str_from_utf8(before).is_ok() {
+                    return;
+                }
+            }
+
+            panic!("byte index {index} is not an OsStr boundary");
+        }
+
+        let encoded_bytes = self.as_encoded_bytes();
+        let Range { start, end } = slice::range(range, ..encoded_bytes.len());
+        check_valid_boundary(encoded_bytes, start);
+        check_valid_boundary(encoded_bytes, end);
+
+        // SAFETY: `slice::range` ensures that `start` and `end` are valid
+        let slice = unsafe { encoded_bytes.get_unchecked(start..end) };
+
+        // SAFETY: `slice` comes from `self` and we validated the boundaries
+        unsafe { Self::from_encoded_bytes_unchecked(slice) }
     }
 
     /// Converts this string to its ASCII lower case equivalent in-place.

--- a/library/std/src/ffi/os_str/tests.rs
+++ b/library/std/src/ffi/os_str/tests.rs
@@ -177,3 +177,53 @@ fn into_rc() {
     assert_eq!(&*rc2, os_str);
     assert_eq!(&*arc2, os_str);
 }
+
+#[test]
+fn slice_encoded_bytes() {
+    let os_str = OsStr::new("123θგ🦀");
+    // ASCII
+    let digits = os_str.slice_encoded_bytes(..3);
+    assert_eq!(digits, "123");
+    let three = os_str.slice_encoded_bytes(2..3);
+    assert_eq!(three, "3");
+    // 2-byte UTF-8
+    let theta = os_str.slice_encoded_bytes(3..5);
+    assert_eq!(theta, "θ");
+    // 3-byte UTF-8
+    let gani = os_str.slice_encoded_bytes(5..8);
+    assert_eq!(gani, "გ");
+    // 4-byte UTF-8
+    let crab = os_str.slice_encoded_bytes(8..);
+    assert_eq!(crab, "🦀");
+}
+
+#[test]
+#[should_panic(expected = "byte index 2 is not an OsStr boundary")]
+fn slice_mid_char() {
+    let crab = OsStr::new("🦀");
+    let _ = crab.slice_encoded_bytes(..2);
+}
+
+#[cfg(windows)]
+#[test]
+#[should_panic(expected = "byte index 3 is not an OsStr boundary")]
+fn slice_between_surrogates() {
+    use crate::os::windows::ffi::OsStringExt;
+
+    let os_string = OsString::from_wide(&[0xD800, 0xD800]);
+    assert_eq!(os_string.as_encoded_bytes(), &[0xED, 0xA0, 0x80, 0xED, 0xA0, 0x80]);
+    let _ = os_string.slice_encoded_bytes(..3);
+}
+
+#[cfg(windows)]
+#[test]
+fn slice_surrogate_edge() {
+    use crate::os::windows::ffi::OsStringExt;
+
+    let os_string = OsString::from_wide(&[0xD800]);
+    let mut with_crab = os_string.clone();
+    with_crab.push("🦀");
+
+    assert_eq!(with_crab.slice_encoded_bytes(..3), os_string);
+    assert_eq!(with_crab.slice_encoded_bytes(3..), "🦀");
+}

--- a/library/std/src/lib.rs
+++ b/library/std/src/lib.rs
@@ -341,6 +341,7 @@
 #![feature(round_ties_even)]
 #![feature(slice_internals)]
 #![feature(slice_ptr_get)]
+#![feature(slice_range)]
 #![feature(std_internals)]
 #![feature(str_internals)]
 #![feature(strict_provenance)]

